torchnlp/metrics/rouge.py

```diff
@@ -0,0 +1,332 @@
+        eps: error threshold
+    Returns:
+        bool, if inv_f is the inverse function of f
+    """
+    for i in range(100000):
+        x = random.random() * 100000
+        y = f(x)
+        _x = inv_f(y)
+        if abs(x - _x) > eps:
+            return False
+    return True
+
+
+def check_increase(f):
+    """
+    Check if function f satisfies f(x + y) > f(x) + f(y)
+    Args:
+        f: function
+    Returns:
+        bool, if f(x + y) > f(x) + f(y) or not
+    """
+    for i in range(100000):
+        x = random.random() * 100000
+        y = random.random() * 100000
+        if f(x) + f(y) <= f(x + y):
+            return False
+    return True
+
+
+def _frp_rouge_w(wlcs, m, n, f=lambda x: x**2, inv_f=lambda x: math.sqrt(x),
+                 beta=None, strict=False):
+    """
+    Computes the LCS-based F-measure score.
+    Args:
+       wlcs: wlcs score
+       m: number of words in reference summary
+       n: number of words in candidate summary
+       f: weighting function
+       inv_f: inverse function of weighting function
+       beta: beta = P_lcs / R_lcs when ∂ F_lcs / ∂ R_lcs = ∂ F_lcs / ∂ P_lcs.
+           In DUC, beta is set to a very big number, and only R_lcs is considered.
+    Returns:
+       dictionary. WLCS-based F-measure score, P-score and R-score
+    """
+    if strict:
+        assert(check_increase(f))
+        assert(check_inverse(f, inv_f))
+    r_wlcs = inv_f(wlcs / f(m))
+    p_wlcs = inv_f(wlcs / f(n))
+    if not beta:
+        beta = p_wlcs / (r_wlcs + 1e-12)
+    num = (1 + (beta**2)) * r_wlcs * p_wlcs
+    denom = r_wlcs + ((beta**2) * p_wlcs)
+    f_wlcs = num / (denom + 1e-12)
+    return {"f": f_wlcs, "p": p_wlcs, "r": r_wlcs}
+
+
+def lcs(x, y):
+    """
+    Compute the length of longest common sequence of x and y
+    Args:
+        x, y: List of element
+    Returns:
+        Dictionary, table[i, j] represent the length of
+            the longest common sequence of x[0: i-1] and y[0: j-1]
+    """
+    n, m = len(x), len(y)
+    table = {}
+    for i in range(n + 1):
+        for j in range(m + 1):
+            if i == 0 or j == 0:
+                table[i, j] = 0
+            elif x[i - 1] == y[j - 1]:
+                table[i, j] = table[i - 1, j - 1] + 1
+            else:
+                table[i, j] = max(table[i - 1, j], table[i, j - 1])
+    return table
+
+
+def lcs_seq(x, y):
+    """
+    Find the longest common sequence of x and y
+    Args:
+        x, y: List of element
+    Returns:
+        List, the longest common sequence of x and y
+    """
+    table = lcs(x, y)
+
+    def lcs_seq_wrd(i, j):
+        if i == 0 or j == 0:
+            return []
+        if x[i - 1] == y[j - 1]:
+            return lcs_seq_wrd(i - 1, j - 1) + [x[i - 1]]
+        elif table[i - 1, j] > table[i, j - 1]:
+            return lcs_seq_wrd(i - 1, j)
+        else:
+            return lcs_seq_wrd(i, j - 1)
+    return lcs_seq_wrd(len(x), len(y))
+
+
+def _w_lcs(x, y, func=lambda x: x**2):
```

Comment:
This algorithm is fairly complex. Can you provide a full test suite to make sure it's correct?